tox.ini

```diff
@@ -43,3 +43,10 @@ max_line_length = 79
 statistics = True
 ignore = E121,E122,E123,E125,E126,E127,E128,E226,E231,E251,E265,E266,E302,E402,E501,E712,E721,E731,E741,W291,W293,W391,W503,W504
 exclude = scipy/__config__.py
+
+[flake8]
+max_line_length = 79
+ignore = E121,E122,E123,E125,E126,E127,E128,E226,E231,E251,E265,E266,E302,E402,E501,E712,E721,E731,E741,
+         W291,W293,W391,W503,W504,
+         F401,F403,F405,F841
```

Comment:
The E and W codes are the same as for `pycodestyle`, I've just broken them onto separate lines and added these F codes.

Reply:
Why not add `F822`? Otherwise, all the deprecated modules proposed in #14360 will have to add the ` # noqa: F822` comment. If it is such a common occurrence, should we ignore it globally?

Reply:
In most other situations it would indicate a bug. Let's now throw out the baby with the bath water.

Reply:
@peterbell10 @rgommers These rules appear to be ignored locally, but it seems they are not being ignored on CI. 

For example, in gh-14433, Main Lint failed due to:

```
benchmarks/benchmarks/sparse_linalg_svds.py:2:1: F401 'numpy as np' imported but unused
scipy/sparse/linalg/__init__.py:128:1: F403 'from ._svdp import *' used; unable to detect undefined names
scipy/sparse/linalg/__init__.py:128:1: F401 '._svdp.*' imported but unused
scipy/sparse/linalg/_propack/setup.py:20:5: F401 'scipy._build_utils.get_g77_abi_wrappers' imported but unused
scipy/sparse/linalg/_propack/setup.py:26:5: E265 block comment should start with '# '
scipy/sparse/linalg/_svdp.py:50:1: E302 expected 2 blank lines, found 1
scipy/sparse/linalg/_svdp.py:82:10: E128 continuation line under-indented for visual indent
scipy/sparse/linalg/_svdp.py:83:10: E128 continuation line under-indented for visual indent
scipy/sparse/linalg/_svdp.py:84:10: E128 continuation line under-indented for visual indent
scipy/sparse/linalg/eigen/_svds_doc.py:256:22: E128 continuation line under-indented for visual indent
scipy/sparse/linalg/eigen/_svds_doc.py:257:22: E128 continuation line under-indented for visual indent
scipy/sparse/linalg/eigen/tests/test_svds.py:422:31: E128 continuation line under-indented for visual indent
scipy/sparse/linalg/eigen/tests/test_svds.py:573:1: E302 expected 2 blank lines, found 1
scipy/sparse/linalg/eigen/tests/test_svds.py:580:1: E302 expected 2 blank lines, found 1
scipy/sparse/linalg/eigen/tests/test_svds.py:614:1: E302 expected 2 blank lines, found 1
scipy/sparse/linalg/eigen/tests/test_svds.py:622:1: E302 expected 2 blank lines, found 1
scipy/sparse/linalg/setup.py:15:1: W293 blank line contains whitespace
scipy/sparse/linalg/tests/test_propack.py:48:5: E741 ambiguous variable name 'I'
scipy/sparse/linalg/tests/test_propack.py:243:5: F841 local variable 'dtype' is assigned to but never used
```

yet now when I run `flake8` locally, issues like F401 and E128 don't show up. This seems opposite the way I would like - I don't want these things ignored locally, as I'd like to be warned of things like unneeded imports, but it's OK with me if CI is less strict. Am I doing something wrong?

Reply:
Running `flake8` applies a limited set of rules that are known to pass for the whole code base, whereas `lint_diff.py` applies a stricter set to the changes in your PR. This was also how it worked before this PR.

Reply:
>  It is going to be really hard to get CI passing if flake8 doesn't report errors locally that it does on CI.

@mdhaber you can check if CI will pass or fail by running `python tools/lint_diff.py` on your feat branch.

AFAIK the CI is checking linting only on the diff which is governed by [`lint_diff.py`](https://github.com/scipy/scipy/blob/master/tools/lint_diff.py). From what I could understand, we don't want to fix the lint failures existing in master (maybe git blame reasons) already so we ignore them (less strict), but we also don't want to introduce new linting issues (hence more strict in PRs/CI).

Peter can probably confirm this if I'm not wrong.

Reply:
Yes, the motivation is that new code follows a clean PEP8 style and old code doesn't need to be changed all at once.

Reply:
I've never used `lint_diff.py`. I've always used `flake8` directly. `lint_diff.py` sounds great, now that I know about it. Please consider documenting it [here](http://scipy.github.io/devdocs/dev/contributor/pep8.html).

I ran `python lint_diff.py` on my feature branch and it lists hundreds of issues unrelated to the PR. It sounds like that's not how it's supposed to work. In the meantime, other than deleting the new `flake8` rule exceptions, is there a way to locally check the failures just for the files I'm interested in?

Reply:
Oh, I think something went wrong in the PR. When we fix that, I imagine `lint_diff.py` will work correctly.

Reply:
You can always pass a config file to flake8 so that you test your changes.
```bash
flake8 --config tools/lint_diff.ini scipy/sparse/linalg/tests/test_propack.py
```
But this is not a complete solution. In the case when you edit an existing file that has flake8 issues (but were suppressed in master in the past due to `tox.ini`) and then you make some changes adding new flake8 errors to that same file. The way I mention will throw out old and both new errors at once.

Anyways it's important we update the contributor documentation. I'll send a PR to fix that.

At the same time, I'll add a small patch in the PR to offer an argument to pass in files you want to run `lint_diff` on specific files only. Something like below should be possible then.
```bash
python tools/lint_diff.py --files scipy/sparse/linalg/eigen/_svds_doc.py
>>>scipy/sparse/linalg/eigen/_svds_doc.py:256:22: E128 continuation line under-indented for visual indent
>>>scipy/sparse/linalg/eigen/_svds_doc.py:257:22: E128 continuation line under-indented for visual indent
```

@mdhaber please let me know if this sounds good to you?

Reply:
Yes, that sounds great. Thanks!